libzypp (and anything using it) fails to build with clang due to std::mem_fn

libzypp and anything using libzypp's `<zypp/Arch.h>` header fails to build with clang because clang is stricter about `std::mem_fn` than gcc, rejecting `std::mem_fn` use on a virtual member.

```sh
/home/bero/abf/libzypp/BUILD/libzypp-17.31.18/zypp/Arch.h:137:64: error: no matching function for call to 'mem_fn'
      return str::join( make_transform_iterator( cset.begin(), std::mem_fn(&Arch::asString) ),
                                                               ^~~~~~~~~~~
/usr/bin/../lib64/gcc/x86_64-openmandriva-linux-gnu/13.1.1/../../../../include/c++/13.1.1/functional:237:5: note: candidate template ignored: couldn't infer template argument '_Tp'
    mem_fn(_Tp _Class::* __pm) noexcept
    ^
```

Thanks, that does fix it. libzypp itself still doesn't build with clang (at least clang 16) because

```
/home/bero/abf/libzypp/BUILD/libzypp-17.31.18/zypp-tui/output/Out.h:450:36: error: integer value 255 is outside the valid range of values [0, 3] for this enumeration type [-Wenum-constexpr-conversion]
  static constexpr Type TYPE_ALL        = TypeBit(0xff);
                                          ^
1 warning and 1 error generated.
```

Clang realizes the enum only takes values of 0 to 3, so doesn't like the 0xff. Setting it to 3 (which for now should do the same thing) fixes it.

--- a/zypp-tui/output/Out.h
+++ b/zypp-tui/output/Out.h
@@ -446,8 +446,8 @@ public:
   };
   ZYPP_DECLARE_FLAGS(Type,TypeBit);
 
-  static constexpr Type TYPE_NONE	= TypeBit(0x00);
-  static constexpr Type TYPE_ALL	= TypeBit(0xff);
+  static constexpr Type TYPE_NONE       = Type(0x00);
+  static constexpr Type TYPE_ALL	= Type(0xff);
 
   using PromptId = unsigned;
 

--- a/zypp/Arch.h
+++ b/zypp/Arch.h
@@ -134,8 +134,11 @@ namespace zypp
     /** */
     static std::string asString( const CompatSet & cset )
     {
-      return str::join( make_transform_iterator( cset.begin(), std::mem_fn(&Arch::asString) ),
-                        make_transform_iterator( cset.end(), std::mem_fn(&Arch::asString) ) );
+      // Explicitely select the overload we want, clang seems to choke if there is a static overload for the
+      // member function we want to capture in std::mem_fn (fixes GH-478)
+      const std::string & (Arch::*memfn)( ) const = &Arch::asString;
+      return str::join( make_transform_iterator( cset.begin(), std::mem_fn( memfn ) ),
+                        make_transform_iterator( cset.end(), std::mem_fn( memfn ) ) );
     }
 
   public:

